File: printer_server/drivers/screen/screen_snip.js
var enable_upload_button = function (le) {
    $(`#${le}-upload-btn`).prop('disabled', false);
    $(`#${le}-upload-btn`).addClass('btn-outline-info');
    $(`#${le}-upload-btn`).removeClass('btn-outline-secondary');
}

var disable_upload_button = function (le) {
    $(`#${le}-upload-btn`).prop('disabled', true);
    $(`#${le}-upload-btn`).removeClass('btn-outline-info');
    $(`#${le}-upload-btn`).addClass('btn-outline-secondary');
}

var highlight_draw_button = function (le) {
    $(`#${le}-draw-btn`).addClass('active');
}

var unhighlight_draw_button = function (le) {
    $(`#${le}-draw-btn`).removeClass('active');
}

var highlight_draw_white_button = function (le) {
    $(`#${le}-white-btn`).addClass('active');
}

var unhighlight_draw_white_button = function (le) {
    $(`#${le}-white-btn`).removeClass('active');
}

$(document).ready(function () {
    // Once once upload is complete, re-enable upload controls
    socket.on("screen_image_uploaded", function (le) {
        let filePickerElement = document.getElementById(`${le}-file-picker`);
        filePickerElement.classList.remove("is-invalid")
        // enable_upload_button(le);
    });

    // If a bad file was uploaded, disable upload options
    socket.on("screen_image_bad", function (le) {
        let filePickerElement = document.getElementById(`${le}-file-picker`);
        filePickerElement.classList.add("is-invalid")
        enable_upload_button(le);
    });

    socket.on('screen_load', function(message) {
        for (let le in manual_controls_data["light_engines"]) {
            let lightCheckboxElement = document.getElementById(`${le}-light-correction-chkbx`);
            let darkCheckboxElement = document.getElementById(`${le}-dark-correction-chkbx`);
            lightCheckboxElement.checked = message[le]["light"];
            darkCheckboxElement.checked = message[le]["dark"];
            
        }
    });

    socket.on('screen_previews', function(message) {
        for (let le in manual_controls_data["light_engines"]) {
            var img = document.getElementById(`${le}-preview`);
            if (message[le]) {
                img.src = 'data:image/jpeg;base64,' + message[le];
            }
            img.style.display = 'inline';
        }
    });

    socket.on('screen_done', function(message) {
        for (const [le, data] of Object.entries(message)) {
            var img = document.getElementById(`${le}-preview`);
            if (data) {
                img.src = 'data:image/jpeg;base64,' + data;
            }
            img.style.display = 'inline';
        }
    });


    for (let le in manual_controls_data["light_engines"]) {

        disable_upload_button(le);

        document.getElementById(`${le}-file-picker`).addEventListener('change', function (event) {
            let projector = $(this).closest(".container").attr('aria-label');
            filePickerElement = event.currentTarget
            const curFiles = filePickerElement.files;
            if (curFiles.length === 0) {
                disable_upload_button(projector);
            } else {
                enable_upload_button(projector);
            }
        });

        // Correction button click function
        $(`#${le}-light-correction-chkbx`).on("click", function (e) {
            let projector = $(this).closest(".container").attr('aria-label');
            let checkboxElement = document.getElementById(`${le}-light-correction-chkbx`);
            let correction = Number(checkboxElement.checked);
            socket.emit("screen_light_grayscale_correction", { "light_engine": projector, "correction":  correction});
        });

        // Correction button click function
        $(`#${le}-dark-correction-chkbx`).on("click", function (e) {
            let projector = $(this).closest(".container").attr('aria-label');
            let checkboxElement = document.getElementById(`${le}-dark-correction-chkbx`);
            let correction = Number(checkboxElement.checked);
            socket.emit("screen_dark_grayscale_correction", { "light_engine": projector, "correction":  correction});
        });

        // Upload button click function
        $(`#${le}-upload-btn`).on("click", function (e) {
            let projector = $(this).closest(".container").attr('aria-label');
            let filePickerElement = document.getElementById(`${projector}-file-picker`);
            let selectedFile = filePickerElement.files[0];
            if (typeof selectedFile !== 'undefined') { // if there is a file selected
                uploadFile(selectedFile, projector);
                disable_upload_button(projector);
                unhighlight_draw_button(projector);
            }
        });

        // Draw button click function
        $(`#${le}-white-btn`).on("click", function (e) {
            let projector = $(this).closest(".container").attr('aria-label');
            socket.emit("screen_white", { "light_engine": projector });
            highlight_draw_white_button(projector);
            unhighlight_draw_button(projector);
        });

        // Draw button click function
        $(`#${le}-draw-btn`).on("click", function (e) {
            let projector = $(this).closest(".container").attr('aria-label');
            socket.emit("screen_draw", { "light_engine": projector });
            highlight_draw_button(projector);
            unhighlight_draw_white_button(projector);
        });

        // Clear button click function
        $(`#${le}-clear-btn`).on("click", function (e) {
            let projector = $(this).closest(".container").attr('aria-label');
            socket.emit("screen_clear", { "light_engine": projector });
            unhighlight_draw_button(projector);
            unhighlight_draw_white_button(projector);
        });
    };
});

function uploadFile(image, le) {
    let fd = new FormData(); // Create form data
    fd.append("file", image); // Attach the image file
    fd.append("light_engine", le);
    console.log(le)
    $.ajax({ // Use ajax to compose and send the request
        url: "/screen_image_upload",
        method: "POST",
        contentType: false,
        processData: false,
        cache: false,
        data: fd
    });
}

File: printer_server/drivers/screen/screen_snip.py
import os
import logging
from printer_server.settings import Config
from printer_server.extensions import socketio
from flask import request, Blueprint, render_template
from printer_server.hardware_configuration.hardware_configuration import driver_handles, config_dict

from PIL import Image

screen = driver_handles.screen
light_engines = driver_handles.light_engines

log = logging.getLogger(__name__)
log.setLevel(logging.INFO)

def handleUpload(request):
    try:
        if "file" in request.files:  # Check if the post request has the file part
            file = request.files["file"]  # Get the file
            light_engine = request.form["light_engine"]
            # Specify location of uploaded image and give default name
            imagePath = os.path.join(
                Config.UPLOAD_FOLDER, "calibration_images", f"{light_engine}.png"
            )
            if file.filename != "" and file:  # File part of request actually has a file
                try:
                    with Image.open(file) as img:  # Open file as PIL object
                        # Check imagePath format and mode
                        if img.format == "PNG" and img.mode == "L":
                            # Seek to the beginning of file (fixes bug in Werkzeug file I\O)
                            file.stream.seek(0)
                            file.save(imagePath)  # save it to the server
                            socketio.emit(
                                "screen_image_uploaded",
                                light_engine,
                                namespace="/manual"
                            )
                            return ""
                except (OSError, FileNotFoundError):  # File has big issues
                    pass
        socketio.emit(
            "screen_image_bad", light_engine, namespace="/manual"
        )
        return ""
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")

def screenLoad():
    try:
        data = {}
        for light_engine in screen.light_engines:
            light_corrected = screen.getLightCorrectionEnable(light_engine)
            dark_corrected = screen.getDarkCorrectionEnable(light_engine)
            data[light_engine] = {
                "light": light_corrected, 
                "dark": dark_corrected
            }
        socketio.emit(
            "screen_load", data, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")


def screenFetchPreviews():
    try:
        previews = {}
        for light_engine in screen.light_engines:
            previews[light_engine] = screen.fetch_preview(light_engine)
        socketio.emit(
            "screen_previews", previews, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")


@socketio.on("screen_light_grayscale_correction", namespace="/manual")
def screenLightGrayscaleCorrection(message):
    try:
        light_engine = message["light_engine"]
        correction = bool(message["correction"])
        screen.setCorrectionEnable(correction, screen.getDarkCorrectionEnable(light_engine), light_engine=light_engine)
        socketio.emit(
            "screen_done", {light_engine: screen.fetch_preview(light_engine)}, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")

@socketio.on("screen_dark_grayscale_correction", namespace="/manual")
def screenDarkGrayscaleCorrection(message):
    try:
        light_engine = message["light_engine"]
        correction = bool(message["correction"])
        
        screen.setCorrectionEnable(screen.getLightCorrectionEnable(light_engine), correction, light_engine=light_engine)
        socketio.emit(
            "screen_done", {light_engine: screen.fetch_preview(light_engine)}, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")

@socketio.on("screen_draw", namespace="/manual")
def screenDraw(message):
    try:
        light_engine = message["light_engine"]
        led_num = light_engines[light_engine].getCurrentLed()
        imagePath = os.path.join(
            Config.UPLOAD_FOLDER, "calibration_images", f"{light_engine}.png"
        )
        screen.draw(imagePath, light_engine=light_engine, led_num=led_num)
        socketio.emit(
            "screen_done", {light_engine: screen.fetch_preview(light_engine)}, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")


@socketio.on("screen_white", namespace="/manual")
def screenWhite(message):
    try:
        light_engine = message["light_engine"]
        led_num = light_engines[light_engine].getCurrentLed()
        imagePath = os.path.join(
            Config.PRINT_SERVER_FOLDER, f"drivers/{light_engine}/images", f"white.png"
        )
        screen.draw(imagePath, light_engine=light_engine, led_num=led_num)
        socketio.emit(
            "screen_done", {light_engine: screen.fetch_preview(light_engine)}, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")


@socketio.on("screen_clear", namespace="/manual")
def screenClear(message):
    try:
        light_engine = message["light_engine"]
        screen.clear(light_engine)
        socketio.emit(
            "screen_done", {light_engine: screen.fetch_preview(light_engine)}, namespace="/manual"
        )
    except Exception as ex:
        log.warn("Screen manual control failed (%s)", ex, exc_info=True)
        socketio.emit("hardware_failure", "screen", namespace="/manual")
